### PR TITLE
Install all versions of .Net Core on ubuntu1804

### PR DESCRIPTION
**Pull request by bishal-pdMSFT**

### CODE REVIEW

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 1]
```diff
@@ -8,39 +8,89 @@
...
-dotnet new globaljson --sdk-version "2.1.105" --force
-dotnet new "mstest"
-dotnet restore
-dotnet build
-set +e
-cd ..
-rm -rf 2.1.105
\ No newline at end of file
+echo "Determing if .NET Core ($LATEST_DOTNET_PACKAGE) is installed"
+if ! IsInstalled $LATEST_DOTNET_PACKAGE; then
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Should we kill this logic now that you are installing all versions

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 2]
```diff
@@ -8,39 +8,89 @@
...
+release_urls=("https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/2.1/releases.json" "https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/2.2/releases.json")
+sdks=()
+for release_url in ${release_urls[@]}; do
+    echo "${release_url}"
+    releases=$(curl "${release_url}")
+    sdks=("${sdks[@]}" $(echo "${releases}" | jq '.releases[]' | jq '.sdk.version'))
+    sdks=("${sdks[@]}" $(echo "${releases}" | jq '.releases[]' | jq '.sdks[]?' | jq '.version'))
+done
+
+sortedSdks=$(echo ${sdks[@]} | tr ' ' '\n' | grep -v preview | grep -v rc | grep -v display | cut -d\" -f2 | sort -u)
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: sorting is not really required

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 3]
```diff
@@ -8,39 +8,89 @@
...
+
+for sdk in $sortedSdks; do
+    echo "$sdk"
+    # Glob matches dotnet-dev-1.x or dotnet-sdk-2.y
+    if ! apt-get install -y --no-install-recommends "dotnet-*-$sdk"; then
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Should we only rely on the download URLs

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 4]
```diff
@@ -8,39 +8,89 @@
...
+    mksamples "$sdk" "webapi"
+    DocumentInstalledItem ".NET Core SDK $sdk"
+done
+
+# NuGetFallbackFolder at /usr/share/dotnet/sdk/NuGetFallbackFolder is warmed up by smoke test
+# Additional FTE will just copy to ~/.dotnet/NuGet which provides no benefit on a fungible machine
+echo "DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1" | tee -a /etc/environment
+echo "PATH=\"/home/vsts/.dotnet/tools:$PATH\"" | tee -a /etc/environment
+
+dotnet --info
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: why is this required

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 1 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: This ensures that dotnet.exe is from latest version

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 3 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: done

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 4 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: removed

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 2 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Sorting will help with documentation order. I have reversed the sorting order so that latest version comes first in doc.

**Comment by bishal-pdMSFT** [comment]
@pasangh .Net core seems to be working but PS core installation fails with these changes. I am updating the PS Core steps according to [here](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux?view=powershell-6#ubuntu-1804) to see if it works.

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 5]
```diff
@@ -8,39 +8,83 @@
...
+    rsync -qav "$dest/sdk/" /usr/share/dotnet/sdk/
+    rm -rf "$dest"
+    rm "$tarball"
+done
+rm urls
+
+DocumentInstalledItem ".NET Core SDK:"
+# Smoke test each SDK
+for sdk in $sortedSdks; do
+    mksamples "$sdk" "console"
```
Comment (pasangh) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Do we require all these mksamples example - has it caught anything in the past? 

It'll be interesting to know the end to end running time of the entire script. Since we are installing and testing all the versions, I think it'll be good to check if we can cut down anything during Image generation. Just a thought :)

**Comment by bishal-pdMSFT** [comment]
> @pasangh .Net core seems to be working but PS core installation fails with these changes. I am updating the PS Core steps according to [here](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux?view=powershell-6#ubuntu-1804) to see if it works.

Here is the failed build: https://dev.azure.com/mseng/AzureDevOps/_build/results?buildId=10015770&view=logs&j=1b9d3bb2-c75f-5891-8836-f664b6fcadfc

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 6]
```diff
@@ -8,39 +8,88 @@
...
-rm -rf 2.1.105
\ No newline at end of file
+echo "Determing if .NET Core ($LATEST_DOTNET_PACKAGE) is installed"
+if ! IsInstalled $LATEST_DOTNET_PACKAGE; then
+    echo "Could not find .NET Core ($LATEST_DOTNET_PACKAGE), installing..."
+    curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg
+    mv microsoft.gpg /etc/apt/trusted.gpg.d/microsoft.gpg
+    sh -c 'echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod bionic main" > /etc/apt/sources.list.d/dotnetdev.list'
+    apt-get install apt-transport-https
+    apt-get update
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: dont we need sudo now for these commands

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh [hunk 7]
```diff
@@ -8,39 +8,88 @@
...
+
+for sdk in $sortedSdks; do
+    url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$sdk/dotnet-sdk-$sdk-linux-x64.tar.gz"
+    echo "$url" >> urls
+    echo "Adding $url to list to download later"
+done
+
+# Download additional SDKs
+echo "Downloading release tarballs..."
+cat urls | xargs -n 1 -P 16 wget -q
```
Comment (sachinma) [inline images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Instead of downloading all togther. should you process one version at a time

**Comment by bishal-pdMSFT** [comment]
Verified that BuildCanary Asp.Net Core build is green: https://buildcanary.visualstudio.com/CanaryBuilds/_build/results?buildId=532339&view=results

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 6 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: No, packer runs this script as sudo

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 7 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Currently they get downloaded in parallel so it saves time

File: images/linux/scripts/installers/1804/dotnetcore-sdk.sh (hunk 5 above)
Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Not only in catching errors, this also helps in pre-warming the nuget store. So, when user runs his job it will be faster as nuget packages are already up to date.

Comment (bishal-pdMSFT) [reply images/linux/scripts/installers/1804/dotnetcore-sdk.sh]: Another reason is that if we don't do this, "First time experience" of .Net Core is shown during pipeline job which can cause some issues (we had one customer issue related to assembly locking in past)
